### PR TITLE
Refactor of contract profiler

### PR DESCRIPTION
**Pull request by MaksymilianDemitraszek**
Transaction profiler will be refactored in a separeate PR. 

This PR adds:
- some docstring
- some structure simplifications
- improves typing
- adds some test for functions which are unit-testable

I will add general integration tests after I integrate profiler in the final way (we discussed this integration with @kasperski95), right now those are a litlle too difficult to setup.

#828

### CODE REVIEW

File: protostar/profiler/contract_profiler.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+)
+from starkware.cairo.lang.vm.memory_segments import MemorySegmentManager
+
+
+Address = int
+StringID = int
+FunctionID = int
+
+
+@dataclass()
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:23]:
```suggestion
@dataclass
```

File: protostar/profiler/contract_profiler.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+class Function:
+    id: str
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:25]:
```suggestion
    id: FunctionID
```

File: protostar/profiler/contract_profiler.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    instructions: List[Instruction]
+    step_samples: List[Sample]
+    memhole_samples: List[Sample]
+    contract_call_callstacks: List[List[Instruction]]
+
+
+class ProfilerContext:
+    def __init__(self, initial_fp: Address, memory: MemoryDict):
+        self.initial_fp = initial_fp
+        self.memory = memory
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:62]:
```suggestion
        self._initial_fp = initial_fp
        self._memory = memory
```

**Review by kasperski95 — APPROVED** [review]

File: protostar/profiler/contract_profiler.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
+# pylint: disable=invalid-name
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:1]: Actually StringID does not exist in this layer anymore, good catch

File: protostar/profiler/contract_profiler.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        for trace_entry in tracer_data.trace:
+            callstack = self.get_call_stack(fp=trace_entry.fp, pc=trace_entry.pc)
+            instr_callstack = [
+                self.find_instruction(instructions, pc) for pc in callstack
+            ]
+            step_samples.append((Sample(value=1, callstack=instr_callstack)))
+        return step_samples
+
+    @staticmethod
+    def blame_pc(last_accesses: Dict[Address, Address], hole_address: Address) -> int:
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:199]: `blame` is a verb

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:199]: This functions blames pc, bot finds blamed. No pc is blamed beforehand

File: protostar/profiler/contract_profiler.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+            if func.id == function_id:
+                return func
+        assert False
+
+    def build_instructions_map(
+        self, functions: List[Function], tracer_data: TracerData
+    ) -> List[Instruction]:
+        """
+        Builds a list of instructions in the contract and assings them to functions
+        """
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:134]: Good catch, it was a map before, forgot to change it

File: protostar/profiler/contract_profiler.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        return Function(
+            id="__main__",
+            filename="<dummy_filename>",
+            start_line=0,
+        )
+
+    def build_function_list(self, tracer_data: TracerData) -> List[Function]:
+        """
+        Collects all of the functions in the compiled contract
+        """
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:89]: `collect_contract_functions` would be better, but yeah

File: protostar/profiler/contract_profiler.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    id: int
+    pc: Address
+    function: Function
+    line: int
+
+
+@dataclass(frozen=True)
+class Sample:
+    value: int
+    callstack: List[Instruction]
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:41]: I prefer List here it's more descriptive

File: protostar/profiler/contract_profiler.py (hunk 2 above)
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:25]: Done

File: protostar/profiler/contract_profiler.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/profiler/contract_profiler.py:199]: > blame is a verb

It is. I believe I read it as `blamed_pc`.

> This functions blames pc, not finds blamed one. No pc is blamed beforehand

The doc string is a bit misleading then.

File: protostar/profiler/contract_profiler.py (hunk 3 above)
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:62]: Done

File: protostar/profiler/contract_profiler.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+def build_profile(
+    tracer_data: TracerData,
+    segments: MemorySegmentManager,
+    segment_offsets: Dict[int, int],
+    accessed_memory: Set[RelocatableValue],
+) -> RuntimeProfile:
+    builder = ProfilerContext(
+        initial_fp=tracer_data.trace[0].fp, memory=tracer_data.memory
+    )
+    function_list = builder.build_function_list(tracer_data)
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:289]: Done

File: protostar/profiler/contract_profiler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:89]: Done

File: protostar/profiler/contract_profiler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:1]: I hidden it in `pprof.py` as it is related to pprof format, not our profiling alghoritms

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:1]: Done, also added dataclasses docs

File: protostar/profiler/contract_profiler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py]: it could be named `InstructionCallstack` but I don't like to overuse aliases

**Review by kasperski95 — APPROVED** [review]

**Pull request merged**